src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -404,6 +416,180 @@ internal void CheckConstraintLanguageVersionAndRuntimeSupportForConversion(Synta
+                    Error(diagnostics, ErrorCode.ERR_CollectionLiteralTargetTypeNotConstructible, syntax, targetType);
+                }
+
+                var collectionInitializerAddMethodBinder = this.WithAdditionalFlags(BinderFlags.CollectionInitializerAddMethod);
+                var builder = ArrayBuilder<BoundExpression>.GetInstance(node.Initializers.Length);
+                foreach (var element in node.Initializers)
+                {
+                    var result = BindCollectionInitializerElementAddMethod(
+                        (ExpressionSyntax)element.Syntax,
+                        ImmutableArray.Create(element),
+                        hasEnumerableInitializerType,
+                        collectionInitializerAddMethodBinder,
+                        diagnostics,
+                        implicitReceiver);
+                    result.WasCompilerGenerated = true;
+                    builder.Add(result);
+                }
+                return new BoundCollectionInitializerCollectionLiteralExpression(
+                    syntax,
+                    collectionCreation,
+                    naturalTypeOpt: null, // PROTOTYPE: Support natural type.
+                    wasTargetTyped: true, // PROTOTYPE: Support natural type.
+                    implicitReceiver,
+                    builder.ToImmutableAndFree(),
+                    targetType)
+                { WasCompilerGenerated = wasCompilerGenerated };
+            }
+
+            MethodSymbol? getSpanConstructor(CSharpSyntaxNode syntax, TypeSymbol spanType, WellKnownMember spanMember, BindingDiagnosticBag diagnostics)
+            {
+                return (MethodSymbol?)GetWellKnownTypeMember(spanMember, diagnostics, syntax: syntax)?.SymbolAsMember((NamedTypeSymbol)spanType);
+            }
+
+            bool collectionTypeImplementsIEnumerable(TypeSymbol targetType, CSharpSyntaxNode syntax, BindingDiagnosticBag diagnostics)
+            {
+                // This implementation differs from CollectionInitializerTypeImplementsIEnumerable().
+                // That method checks for an implicit conversion from IEnumerable to the collection type,
+                // but that would allow: Nullable<StructCollection> s = [];
+                var ienumerableType = GetSpecialType(SpecialType.System_Collections_IEnumerable, diagnostics, syntax);
+                var allInterfaces = targetType is TypeParameterSymbol typeParameter
+                    ? typeParameter.AllEffectiveInterfacesNoUseSiteDiagnostics
+                    : targetType.AllInterfacesNoUseSiteDiagnostics;
+                return allInterfaces.Any(t => ienumerableType.Equals(t, TypeCompareKind.AllIgnoreOptions));
+            }
+
+            BoundExpression convertArrayElement(BoundExpression element, TypeSymbol elementType, BindingDiagnosticBag diagnostics)
+            {
+                CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
```

Comment:
I don't think we need to share this between calls to `convertArrayElement()`. This is a `struct` type and since we're already passing in a `BindingDiagnosticBag`, it seems cleaner to have this method update that diagnostic bag than adding a second diagnostic parameter.